e3nn/kernel_mod.py

```diff
@@ -0,0 +1,118 @@
+        parity = 0 (no parity), 1 (even), -1 (odd)
+        '''
+        super().__init__()
+
+        self.Rs_in = SO3.simplifyRs(Rs_in)
+        self.Rs_out = SO3.simplifyRs(Rs_out)
+
+        def filters_with_parity(l_in, p_in, l_out, p_out):
+            nonlocal get_l_filters
+            return [l for l in get_l_filters(l_in, l_out) if p_out == 0 or p_in * (-1) ** l == p_out]
+
+        self.get_l_filters = filters_with_parity
+        self.sh = sh
+
+        assert isinstance(normalization, str), "normalization should be passed as a string value"
+        assert normalization in ['norm', 'component'], "normalization needs to be 'norm' or 'component'"
+        self.normalization = normalization
+
+        def lm_normalization(l_out, l_in):
+            # put 2l_in+1 to keep the norm of the m vector constant
+            # put 2l_ou+1 to keep the variance of each m component constant
+            # sum_m Y_m^2 = (2l+1)/(4pi)  and  norm(Q) = 1  implies that norm(QY) = sqrt(1/4pi)
+            lm_norm = None
+            if normalization == 'norm':
+                lm_norm = math.sqrt(2 * l_in + 1) * math.sqrt(4 * math.pi)
+            elif normalization == 'component':
+                lm_norm = math.sqrt(2 * l_out + 1) * math.sqrt(4 * math.pi)
+            return lm_norm
+
+        # Clebsch-Gordan for filter, input, output
+        # Rs_filter contains all degrees of freedom and L's for filters
+        # paths contains [l_in, l_out, l_filter for each filter channel]
+        Rs_filter, filter_mapping_matrix, paths = SO3.tensor_productRs(Rs_in,
+                                                                      Rs_out,
+                                                                      get_l_output=get_l_filters,
+                                                                      paths=True)
+
+        self.n_path = len(paths)
+
+        # Helper matrix for spherical harmonics
+        Rs_filter_sorted, sort_mix = SO3.sortRs(Rs_filter)
+        Rs_filter_simplify = SO3.simplifyRs(Rs_filter_sorted)
+        irrep_mix = SO3.map_irrep_to_Rs(Rs_filter_simplify)
+        # Contract sort_mix with rep_mix
+        ylm_mix = torch.einsum('ij,il->jl', sort_mix, irrep_mix) 
+        
+
+        # Create and sort mix matrix for radial functions
+        rf_mix = SO3.map_mul_to_Rs(Rs_filter)
+
+        ###########################################
+        #TODO: Write normalization based on paths #
+        ###########################################
+
+        # Create the radial model: R+ -> R^n_path
+        # It contains the learned parameters
+        self.R = RadialModel(self.n_path)
+        self.set_of_l_filters = [L for mul, L, p in Rs_filter_simplify]
+        # Check l_filters and rep_mix have same dim
+        assert sum([2 * L + 1 for L in self.set_of_l_filters]) == irrep_mix.shape[-1]
+
+        # Register mixing matrix buffers
+        self.register_buffer('filter_clebsch_gordan', filter_mapping_matrix)
```

Comment:
maybe these names could be consistent, i.e. both filter_clebsch_gordan or both filter_mapping_matrix

Reply:
Same thing to the names below?

Reply:
done!